### PR TITLE
feat(tracer): Update the TagsList infrastructure to store `int` properties on ITags implementations

### PR DESCRIPTION
**Pull request by zacharycmontoya**
## Summary of changes
Updates the TagsListGenerator and code analyzers to allow ITags implementations to specify a `TagAttribute` with a `int` property, in addition to the existing `string` property support. This allows us to maintain the original `int` primitive so in the future we can emit `int` primitives in our tracing protocols, such as OTLP.

To ensure that all of the code generation and serialization are implemented correctly, this PR also migrates the `http.status_code` tag to this new feature.

## Reason for change
We want to support emitting spans that align with OpenTelemetry semantic conventions. Since span attributes may be specified as integers in the OpenTelemetry spec, we need a mechanism for storing and emitting integers tags. This mechanism allows our internal  integrations to do that by storing a backing `int?` property for those concepts so they can be emitted as integers when emitting OTLP.

### TagsListGenerator changes
`int?` properties are now supported with the following details:
- The generated `GetTag(string key)` now calls `value?.ToString(System.Globalization.CultureInfo.InvariantCulture)` for `int` properties
- The generated `SetTag(string key)` now calls `int.TryParse(value, System.Globalization.NumberStyles.Integer, System.Globalization.CultureInfo.InvariantCulture, out var intValue)` to convert the input string for `int` properties. If the conversion fails or the input was null, the tag is unset
- The generated `EnumerateTags<TProcessor>(ref TProcessor processor)` now emits the key-value pair as a `TagItem<int>` so consumers can customize how they handle integer tags

### ITags changes
- Updates `ITags.EnumerateTags<TProcessor>(ref TProcessor processor)` to require `TProcessor` arguments to implement `IItemProcessor<int>`. Affected classes include
  - `SpanMessagePackFormatter.TagWriter`
  - `OtlpMapper.TagWriter`
  - `TraceFilter.RegexTagFilterProcessor`
  - `TruncatorTagsProcessor`

### HTTP status code updates
- Updates the `IHasStatusCode` interface to require a `int? HttpStatusCode` property
- Updates the `HttpStatusCode` properties to type `int?` for the following classes:
  - `AwsSdkTags`
  - `HttpTags`
  - `InferredProxyTags`
  - `WebTags`
- Replaces calls to `Span.GetTag(Tags.HttpStatusCode)` with a call to a new extension method `Span.GetHttpStatusCode()` that directly returns an `int?`. This affects stats aggregators (which build stats aggregation keys), the NormalizerTraceProcessor (only runs during client-side stats to validate status code tags) and API Security.
- Adds in-line code comments for integration tests that directly assert against `Tags.HttpStatusCode` (i.e. "http.status_code") so we do not lose track of them when introducing HTTP semantics

## Test coverage
To test the code generator, unit tests have been added `TagsListGeneratorTests.cs`.

The existing HTTP and OpenTelemetry SDK integration tests will exercise the updated TagsList implementation by emitting the same tracing data over DD MsgPack and OTLP as before.

## Other details
N/A

### CODE REVIEW

File: tracer/src/Datadog.Trace.SourceGenerators/TagsListGenerator/Sources.cs [hunk 1]
```diff
@@ -163,12 +171,34 @@ public override void SetTag(string key, string? value)
...
                            @""": 
-                    ")
-                      .Append(property.PropertyName)
-                      .Append(
-                           @" = value;
+                    ");
+
+                    if (property.PropertyType is TagListGenerator.PropertyType.NullableInt)
+                    {
+                        sb.Append("if (int.TryParse(value, System.Globalization.NumberStyles.Integer, System.Globalization.CultureInfo.InvariantCulture, out var parsed")
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace.SourceGenerators/TagsListGenerator/Sources.cs:181]: This seems reasonable to me, we should prob add a unit test for it

File: tracer/test/Datadog.Trace.SourceGenerators.Tests/TagsListGeneratorTests.cs [hunk 2]
```diff
@@ -88,6 +88,206 @@ protected override void WriteAdditionalTags(System.Text.StringBuilder sb)
...
+            const string input = @"using Datadog.Trace.SourceGenerators;
+namespace MyTests.TestListNameSpace
+{
+    public class TestList
+    {
+        [Tag(""NameTag"")]
+    	public string Name { get; set; }
+
+        [Tag(""PortTag"")]
+    	public int Port { get; set; }
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/test/Datadog.Trace.SourceGenerators.Tests/TagsListGeneratorTests.cs]: 👍 Needs fixing

File: tracer/src/Datadog.Trace.SourceGenerators/TagsListGenerator/Sources.cs [hunk 3]
```diff
@@ -209,20 +239,41 @@ public override void EnumerateTags<TProcessor>(ref TProcessor processor)
...
                 {
-                    sb.Append(@"if (")
-                      .Append(property.PropertyName)
-                      .Append(@" is not null)
+                    switch (property.PropertyType)
+                    {
+                        case TagListGenerator.PropertyType.NullableInt:
+                            sb.Append(@"if (")
+                              .Append(property.PropertyName)
+                              .Append(@" is not null)
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace.SourceGenerators/TagsListGenerator/Sources.cs:257]: tiny tiny nit I'm not sure about: `HasValue` is more typical with nullables, I'm not sure if we have a sytlistic preference one way or another these days 🤷‍♂️ 
```suggestion
                              .Append(@".HasValue)
```

if so, same tweak applies further down in this file too 🙂

File: tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs [hunk 4]
```diff
@@ -1007,6 +1007,21 @@ public void Process(TagItem<string> item)
                 Count++;
             }
 
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public void Process(TagItem<int> item)
+            {
+                if (item.SerializedKey.IsEmpty)
+                {
+                    _formatter.WriteTag(ref Bytes, ref Offset, item.Key, item.Value.ToString(System.Globalization.CultureInfo.InvariantCulture), _tagProcessors);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs]: I wonder if there's a more efficient API we could use here to avoid the extra `ToString()` 🤔 Not a problem if not or if it's too much hassle, just wondering out loud

File: tracer/src/Datadog.Trace/Agent/TraceSamplers/TraceFilter.cs [hunk 5]
```diff
@@ -261,5 +261,19 @@ public void Process(TagItem<string> item)
...
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Process(TagItem<int> item)
+        {
+            if (!Matched)
+            {
+                if (_filter.KeyPattern.IsMatch(item.Key))
+                {
+                    // Key-only filter: any matching key is sufficient
+                    // Key:Value filter: value must also match
+                    Matched = _filter.ValuePattern is null || _filter.ValuePattern.IsMatch(item.Value.ToString(System.Globalization.CultureInfo.InvariantCulture));
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Agent/TraceSamplers/TraceFilter.cs]: This feels a little annoying to have to do the `ToString()`, but there's not really a better solution I can think of

File: tracer/src/Datadog.Trace/ExtensionMethods/SpanExtensions.cs [hunk 6]
```diff
@@ -101,15 +119,13 @@ internal static void SetHttpStatusCode(this Span span, int statusCode, bool isSe
...
-
             if (span.Tags is IHasStatusCode statusCodeTags)
             {
-                statusCodeTags.HttpStatusCode = statusCodeString;
+                statusCodeTags.HttpStatusCode = statusCode;
             }
             else
             {
-                span.SetTag(Tags.HttpStatusCode, statusCodeString);
+                span.SetTag(Tags.HttpStatusCode, ConvertStatusCodeToString(statusCode));
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ExtensionMethods/SpanExtensions.cs]: This is annoying - convert int to string to then parse it out again 😅 As discussed though, should be edge case, and also we can explore providing a path for it later

File: tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/HttpMessageHandlerTests.cs [hunk 7]
```diff
@@ -106,6 +106,7 @@ public async Task HttpClient_SubmitsTraces(
 
                 foreach (var span in spans)
                 {
+                    // TODO: This lookup may also need to lookup the OTel http.response.status_code
```
Comment (andrewlock) [inline tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/HttpMessageHandlerTests.cs]: Is it worth extracting this lookup to a helper in `TestHelpers` now for this, so you only have to update that helper later when adding support for OTel semantics instead of touching all these test files again?

File: tracer/test/Datadog.Trace.SourceGenerators.Tests/TagsListGeneratorTests.cs [hunk 8]
```diff
@@ -88,6 +88,206 @@ protected override void WriteAdditionalTags(System.Text.StringBuilder sb)
...
+        {
+            const string input = @"using Datadog.Trace.SourceGenerators;
+namespace MyTests.TestListNameSpace
+{
+    public class TestList
+    {
+        [Tag(""TestPort"")]
+    	public int? Id { get; set; }
+    }
+}";
```
Comment (andrewlock) [inline tracer/test/Datadog.Trace.SourceGenerators.Tests/TagsListGeneratorTests.cs]: nit: use raw literals for test input, as it's easier to read, and you can add syntax highlighting iirc e.g.

```suggestion
            //language=csharp
            const string input = 
              """
              using Datadog.Trace.SourceGenerators;
              namespace MyTests.TestListNameSpace
              {
                  public class TestList
                  {
                      [Tag("TestPort")]
                  	public int? Id { get; set; }
                  }
              };
              """
```

**Review by andrewlock — APPROVED** [review]
Looking good to me - a couple of minor questions etc and the AI flagged a couple of genuine bugs I think, but otherwise LGTM

File: tracer/src/Datadog.Trace.SourceGenerators/TagsListGenerator/Sources.cs (hunk 3 above)
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace.SourceGenerators/TagsListGenerator/Sources.cs:257]: I saw that `EnumerateMetrics` also checks if the `double?` property `is not null` so I think we can continue to use this pattern

File: tracer/src/Datadog.Trace/Agent/TraceSamplers/TraceFilter.cs (hunk 5 above)
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/Agent/TraceSamplers/TraceFilter.cs]: My intern Claude suggested an Int to string cache, which is very similar to our previous StatusCodeToString hot path and the Client/Server Error Status arrays. It's here in 775448543e18a87b5e32ac53ab31dbb985b7958c. LMK if it looks ok to you

File: tracer/src/Datadog.Trace/ExtensionMethods/SpanExtensions.cs (hunk 6 above)
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/ExtensionMethods/SpanExtensions.cs]: See 775448543e18a87b5e32ac53ab31dbb985b7958c for a `Datadog.Trace.Util.IntStringCache` solution

File: tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs (hunk 4 above)
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs]: See 775448543e18a87b5e32ac53ab31dbb985b7958c for a `Datadog.Trace.Util.IntStringCache` solution

File: tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/HttpMessageHandlerTests.cs (hunk 7 above)
Comment (zacharycmontoya) [reply tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/HttpMessageHandlerTests.cs]: I added a helper that mimics `SpanExtensions.GetHttpStatusCode` in ea95be730d4539f3a819e1b65b8b14f9be63f470. It might be confusing that it returns a `string` rather than an `int?` but I think it's ok for now?

File: tracer/test/Datadog.Trace.SourceGenerators.Tests/TagsListGeneratorTests.cs (hunk 8 above)
Comment (zacharycmontoya) [reply tracer/test/Datadog.Trace.SourceGenerators.Tests/TagsListGeneratorTests.cs]: Updated to using raw literals in 7e731466eca09bde912f3dab1c2af86bd59f18e4

File: tracer/src/Datadog.Trace.SourceGenerators/TagsListGenerator/Sources.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace.SourceGenerators/TagsListGenerator/Sources.cs:181]: Fixed by setting the value to null if the argument to `SetTag` is null or cannot be parsed to an int in 85c718343f6f65d81cedde60e69f93c1f826b749

File: tracer/test/Datadog.Trace.SourceGenerators.Tests/TagsListGeneratorTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zacharycmontoya) [reply tracer/test/Datadog.Trace.SourceGenerators.Tests/TagsListGeneratorTests.cs]: Whoops, this is now fixed

File: tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs (hunk 4 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs]: Ah, interesting, I was thinking of an API that wrote the int directly to bytes instead of going through the intermediate `string` 🤔 I'm pretty sure .NET (at least newer versions) already has something similar for small (<10) `int`s baked into the runtime. Seems reasonable to me, though I guess it could well be a case of premature optimization on our part 🤷‍♂️

File: tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/HttpMessageHandlerTests.cs (hunk 7 above)
Comment (andrewlock) [reply tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/HttpMessageHandlerTests.cs]: Could rename it to `SpanExtensions.GetHttpStatusCodeAsString()` if you want to disambiguate 🙂 or `SpanExtensions.GetHttpStatusCodeTag()`?

File: tracer/src/Datadog.Trace/OpenTelemetry/OtlpMapper.cs [hunk 9]
```diff
@@ -304,6 +304,32 @@ public void Process(TagItem<string> item)
...
+
+                if (_tagProcessors is not null)
+                {
+                    for (var i = 0; i < _tagProcessors.Length; i++)
+                    {
+                        _tagProcessors[i]?.ProcessMeta(ref key, ref value);
+                    }
+                }
+
+                _writeKeyValue(ref State, new KeyValue(key, value));
```
Comment (bouwkast) [inline tracer/src/Datadog.Trace/OpenTelemetry/OtlpMapper.cs:371]: I think we discussed this (maybe) but how does this interact exactly if the tag came from a manual span? I think this will properly get the autoinstrumentation tag to be the `int`, but I think if we have a manual span that it will potentially still be a `string`?

**Review by bouwkast — APPROVED** [review]
LGTM just a question on how it works with a manual span (assuming someone sets the status code tag)

Which I think was discussed but I forget 😅

File: tracer/src/Datadog.Trace/OpenTelemetry/OtlpMapper.cs (hunk 9 above)
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/OpenTelemetry/OtlpMapper.cs:371]: The manual span doesn't get this because only the internal `Tracer.StartActiveInternal` API allows passing in an instance of `ITags`. When there's no specified `ITags` instance, the `Span` constructor creates a `TagsList` instance, which does not have a field for the well-known status code tag. We could move it up to there since it is such an important tag! But I opted not to for this PR. What's your thought on that?

Comment (bouwkast) [reply tracer/src/Datadog.Trace/OpenTelemetry/OtlpMapper.cs:371]: I think we can just leave it as is thanks!

File: tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/HttpMessageHandlerTests.cs (hunk 7 above)
Comment (zacharycmontoya) [reply tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/HttpMessageHandlerTests.cs]: In the next [PR](https://github.com/DataDog/dd-trace-dotnet/pull/8971) I've renamed this to `SpanExtensions.GetHttpStatusCodeString` 😅

**Review by NachoEchevarria — APPROVED** [review]

**Review by anna-git — APPROVED** [review]
